### PR TITLE
"Breadcrumb" room sorting algorithm

### PR DESCRIPTION
**Pull request by turt2live**
**Caution**: This does not use a labs flag and alters how the room list sorting is done.

*This is being opened now to give people a point to test off of without impacting develop*.

Fixes https://github.com/vector-im/riot-web/issues/8531
Fixes https://github.com/vector-im/riot-web/issues/8122
Fixes https://github.com/vector-im/riot-web/issues/7826
Fixes https://github.com/vector-im/riot-web/issues/7803 ~~(well, not yet...)~~
Fixes https://github.com/vector-im/riot-web/issues/8774

Known bugs:
* [x] Reading rooms on other devices does not cause them to fall down
* [x] ~~UNCONFIRMED:~~ Potentially high CPU usage trying to math all this out
* [x] Direct chats go nuts and end up in the wrong order after refreshing

Room for improvement:
* Tags could be magically calculated rather than bluntly throwing away our incremental lists
* Same applies to direct chat tagging

### CODE REVIEW

**Comment by turt2live** [comment]
Early performance characteristics (based on my account; before optimization):
* Initial room list generation: 0.13ms -> 394ms
* Incremental room list generation: 0.26ms -> 1.5ms

I'm not entirely sure how the old algorithm achieves 0.13ms initial generation time, but it is at least consistent given the same algorithm is used for incremental changes. Suspicion for the new initial list generation taking 400ms is lack of caching, so we'll see how that goes.

File: src/stores/RoomListStore.js [hunk 1]
```diff
@@ -129,63 +164,203 @@ class RoomListStore extends Store {
...
+            let desiredCategoryBoundaryIndex = 0;
+            let foundBoundary = false;
+
+            for (const entry of this._state.lists[key]) {
+                // if the list is a recent list, and the room appears in this list, and we're not looking at a sticky
+                // room (sticky rooms have unreliable categories), try to slot the new room in
+                if (LIST_ORDERS[key] === 'recent' && hasRoom && entry.room.roomId !== this._state.stickyRoomId) {
+                    const inTag = targetTags.length === 0 || targetTags.includes(key);
+                    if (!pushedEntry && doInsert && inTag) {
+                        const entryTimestamp = this._tsOfNewestEvent(entry.room);
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/RoomListStore.js]: All the `false &&` stuff is indeed debugging leftovers - I'll have to verify the time sorting works when the flags are enabled.

It's supposed to sort by timestamp within the category, so more recent red notifications go to the top. This is a bit different from the proposed algorithm due to encrypted events ruining the day for everything.

File: src/stores/RoomListStore.js [hunk 2]
```diff
@@ -129,63 +164,203 @@ class RoomListStore extends Store {
...
+        // interest in trying to maintain the category that they were last in before
+        // being artificially flagged as IDLE. Also, this reduces the amount of time
+        // we spend in _setRoomCategory ever so slightly.
+        if (this._state.stickyRoomId !== room.roomId) {
+            const category = this._calculateCategory(room);
+            this._setRoomCategory(room, category);
+        }
+    }
+
+    _setRoomCategory(room, category) {
```
[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/RoomListStore.js]: tbh I'm not sure how possible that is without long function definitions. With the exception of a few lines, everything directly modifies the collection being iterated on.

File: src/stores/RoomListStore.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bwindels) [reply src/stores/RoomListStore.js]: > It's supposed to sort by timestamp

what timestamp though? I would have thought the time of the notification, not the time of the last event in the room where you got notified?

[the remark this answers is not in the document]
Comment (turt2live) [reply src/stores/RoomListStore.js]: The most recent event timestamp. Sorting by category change gets a bit too magic for me to be comfortable, tbh. We don't have reliable events in place to react to the changes well enough, and fixing that is out of scope of this PR.

[the remark this answers is not in the document]
Comment (bwindels) [reply src/stores/RoomListStore.js]: :+1:

**Comment by turt2live** [comment]
sorry, this technically wasn't back up for review while I'm investigating why its so bad at performance. Intention is to make it transparent, which will probably happen by the time we cut another point release.

**Comment by turt2live** [comment]
I lied: I thought this would take longer to optimize, but I was wrong. After staring at profiles for a few hours and making small changes I've managed to cut it down to the following characteristics:

* First sort time: 2ms worst case, under 1ms most cases.
* Incremental sort time: 1ms worst case, under 0.4ms most cases

**Comment by turt2live** [comment]
ps: executive decision is to merge it asap for testing

**Pull request merged**